### ISSUE TITLE
How we can cancel or pause downloading files using dio?

### ISSUE THREAD

**Issue by Zeeshan0201** [flutterchina/dio#399]
### New Issue Checklist

- [ ] I have searched for a similar issue in the [project](https://github.com/flutterchina/dio/issues) and found none

### Issue Info

| Info                           | Value                                                 |      |
| ------------------------------ | ----------------------------------------------------- | ---- |
| Platform Name                  | e.g. flutter / ios / android                          |      |
| Platform Version               | e.g.  1.5.0 / 12.0 / 9.0                              |      |
| Dio Version                    | e.g.  2.1.0 / 1.0.17                                  |      |
| Android Studio / Xcode Version | e.g. Android Studio 3.3.2 / Xcode 10.2.1              |      |
| Repro rate                     | e.g. all the time (100%) / sometimes x% / only once   |      |
| Repro with our demo prj        | e.g. does it happen with our demo project?            |      |
| Demo project link              | e.g. link to a demo project that highlights the issue |      |

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

**Comment by krypticallusion** [comment]
```CancelToken cancelToken = CancelToken;


dio.download(
    cancelToken: cancelToken,
);```


Now from anywhere just,
    `cancelToken.cancel()`

**Comment by kareemX** [comment]
Have you found a solution?

**Comment by deepashree-ballodi** [comment]
How we can Pause and continue downloading where we stopped exactly?

**Comment by krypticallusion** [comment]
yes

**Comment by 1-alok** [comment]
+1

**Comment by krypticallusion** [comment]
Edited my original comment.

**Comment by Omi231** [comment]
> Edited my original comment.

Can you explain in a bit more detail for noobs like me? Thanks

**Comment by mikezander** [comment]
@krypticallusion 

> EDIT:

Actually, pause/resume also works, but you have to supply the header range:1222-19238, then append the files.
This downloads the file where you left off at (you will have to calculate the downloaded file size)

Can you give a simple example? What do you mean by append the files?

**Comment by krypticallusion** [comment]
> @krypticallusion
> 
> > EDIT:
> 
> Actually, pause/resume also works, but you have to supply the header range:1222-19238, then append the files. This downloads the file where you left off at (you will have to calculate the downloaded file size)
> 
> Can you give a simple example? What do you mean by append the files?

I can't right now give you the exact working sample, sorry.
But here is how it goes :

Z = "The File to be downloaded"

Once you cancel() a download, You should have the file that dio wrote to.
Check the size of the corrupt file.

Issue another download with the range supplied. Suppose you paused the download at 1MB out of 14Mb so (1048576). 
Now while calling the download method, supply `range:1048576-` (No need of the end byte position since you're gonna be downloading the whole thing). Now you have two files, one that has downloaded upto 1048576 bytes while the other file has downloaded from 1048577 to the end of file. Now open the first file with FileMode.APPEND, and write the second file to the first file.

**Comment by krypticallusion** [comment]
https://github.com/flutterchina/dio/blob/develop/example/lib/download_with_trunks.dart

refer this for merging.

**Comment by Ahmed-Sellami** [comment]
Why doesn't the package offer the possibility to use FileMode.append instead of FileMode.write instead of downloading chunks and then merging them?https://github.com/cfug/dio/blob/7cfadc47af1651688c78b6769f8efae55d3d5f11/dio/lib/src/dio/dio_for_native.dart#L89

**Comment by sokryptk** [comment]
> Why doesn't the package offer the possibility to use FileMode.append instead of FileMode.write instead of downloading chunks and then merging them?
> 
> https://github.com/cfug/dio/blob/7cfadc47af1651688c78b6769f8efae55d3d5f11/dio/lib/src/dio/dio_for_native.dart#L89

Beats the point, you can't have parallel file writes at the same point in time at different file locations. That's simply not possible. That's why we create some temporary files that can be parallel-ly written to, then stitch them back together. Hope this helps.

**Comment by Khushnu** [comment]
any solution guys ? for pause and resume ?

**Comment by kostov** [comment]
@Khushnu,
just read the solution above :)
https://github.com/cfug/dio/issues/399#issuecomment-985216178
I also starting to implement this.
